src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, request: DalMethodRequest, path: Optional[Tuple[PathSegment]] = None):
+        self._request = request
+        self._path = path
+
+
+@dataclass
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name: str
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta: dict
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+
+class DalMethodRequester:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+                TODO: Consider using a custom type for the handler to represent
+                      or document the interface to that callable.
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> DalMethodRequester:
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+
+class MethodMiddleware:
+    """
+    Base class used primarily for type checking and to provide
+    an interface for a method middleware.
+    """
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        pass
```

Comment:
- [x] fix

```suggestion
        raise NotImplementedError
```
Otherwise we'd easily forget to implement this in base classes.